### PR TITLE
Fix/no multi team

### PR DESCRIPTION
**Pull request by tbushman**
In response to #49 ... I need help with handling the error but this prevents people from being on multiple teams. Thank you.

### CODE REVIEW

File: server/middlewares/util.js [hunk 1]
```diff
@@ -16,3 +16,31 @@ exports.ifNotOwnTeam400 = function ifNotOwnTeam400(req, res, next) {
   return next();
 };
 
+exports.ifDuplicate400 = function ifDuplicate400(req, res, next) {
```
Comment (ValeraS) [inline server/middlewares/util.js]:
```js
exports.ifDuplicate400 = async function ifDuplicate400(req, res, next) {
  const { body } = req;
  const collaborators = Array.isArray(body.collaborators)
    ? body.collaborators
    : body.collaborators
        .split(',')
        .map(str => str.trim().replace(/@/g, ''))
        .filter(Boolean);
  const filtered = (await Promise.all(collaborators.map(member =>
        Team.find({ collaborators: member })
        .then((dup) => {
          if (dup.length > 0) { return false; }
          return member;
        }))
    )).filter(Boolean);

  if (filtered.length !== collaborators.length) {
    return res.sendStatus(400);
  }
  return next();
};
```

Comment (tbushman) [reply server/middlewares/util.js]: <3

**Comment by tbushman** [comment]
@ValeraS thank you! I am still unclear on how to redirect to '/' without `cannot set headers after they are sent` errors...

**Comment by ValeraS** [comment]
@tbushman How can I reproduce the error?

**Comment by tbushman** [comment]
@ValeraS The page doesn't redirect because when I try to use res.redirect it's occuring after res.error 404 and I get `can't set headers after they are sent`. Thanks for your help. :)

File: server/middlewares/util.js [hunk 2]
```diff
@@ -16,3 +16,25 @@ exports.ifNotOwnTeam400 = function ifNotOwnTeam400(req, res, next) {
   return next();
 };
 
+
+exports.ifDuplicate400 = async function ifDuplicate400(req, res, next) {
+  const { body } = req;
+  const collaborators = Array.isArray(body.collaborators)
```
Comment (Bouncey) [inline server/middlewares/util.js]: The map and filter `.map(str => str.trim().replace(/@/g, '')).filter(Boolean)` still needs to be run over `collaborators` if it is an array

**Review by Bouncey — CHANGES_REQUESTED** [review]

**Comment by tbushman** [comment]
@Bouncey Thanks! @ValeraS I was mistaken on how to handle this. Thanks both for your help.

**Comment by Bouncey** [comment]
LGTM, pulling down to test

**Comment by Bouncey** [comment]
I am able to add myself multiple times, with same casing and different casing. I am also still able to add collaborators that are already collaborating else on different teams.

image

**Comment by tbushman** [comment]
@Bouncey And I'm getting `Unhandled rejection e.exports@https://unpkg.com/axios/dist/axios.min.js:8:4479
e.exports@https://unpkg.com/axios/dist/axios.min.js:8:4321
e.exports/</l[h]@https://unpkg.com/axios/dist/axios.min.js:8:3278` in the browser console. ...

File: server/middlewares/util.js [hunk 3]
```diff
@@ -16,3 +20,22 @@ exports.ifNotOwnTeam400 = function ifNotOwnTeam400(req, res, next) {
   return next();
 };
 
+
+exports.ifDuplicate400 = async function ifDuplicate400(req, res, next) {
+  const { body } = req;
+  const collaborators = castMaybeStringToArray(body.collaborators)
+    .map(str => str.trim().replace(/@/g, ''))
+    .filter(Boolean);
```
Comment (ValeraS) [inline server/middlewares/util.js]: .filter((item, pos, self) => self.indexOf(item) === pos)

File: server/middlewares/util.js [hunk 4]
```diff
@@ -16,3 +20,22 @@ exports.ifNotOwnTeam400 = function ifNotOwnTeam400(req, res, next) {
...
+  const filtered = (await Promise.all(collaborators.map(member =>
+        Team.find({ collaborators: member })
```
Comment (ValeraS) [inline server/middlewares/util.js]: Team.find({ _id: { $ne: req.params.teamId }, collaborators: member })

File: server/middlewares/util.js [hunk 5]
```diff
@@ -16,3 +20,22 @@ exports.ifNotOwnTeam400 = function ifNotOwnTeam400(req, res, next) {
...
+        .then((dup) => {
+          if (dup.length > 0) { return false; }
+          return member;
+        }))
+    )).filter(Boolean);
+
+  if (filtered.length !== collaborators.length) {
+    return res.sendStatus(400);
+  }
+  return next();
```
Comment (ValeraS) [inline server/middlewares/util.js:42]: body.collaborators = collaborators;

File: server/team/team.route.js [hunk 6]
```diff
@@ -13,7 +13,12 @@ router
   /** GET /api/teams - Get list of users */
   .get(teamCtrl.list)
   /** POST /api/teams - Create new team */
-  .post(ifNoUserRedirect(), ifNoBody400, validate(validators.createTeam), teamCtrl.create);
+  .post(
+    ifNoUserRedirect(),
+    ifNoBody400,
+    ifDuplicate400,
```
Comment (ValeraS) [inline server/team/team.route.js:19]: add `ifDuplicate400` to update

**Comment by tbushman** [comment]
@ValeraS Thank you!

File: server/middlewares/util.js [hunk 7]
```diff
@@ -25,9 +25,9 @@ exports.ifDuplicate400 = async function ifDuplicate400(req, res, next) {
   const { body } = req;
   const collaborators = castMaybeStringToArray(body.collaborators)
     .map(str => str.trim().replace(/@/g, ''))
-    .filter(Boolean);
+    .filter((item, pos, self) => self.indexOf(item) === pos);
```
Comment (ValeraS) [inline server/middlewares/util.js]: Sorry, I meant to add not replace.
```js 
.filter((item, pos, self) => !!item && self.indexOf(item) === pos);
```

**Comment by ValeraS** [comment]
> @Bouncey And I'm getting `Unhandled rejection e.exports@https://unpkg.com/axios/dist/axios.min.js:8:4479 e.exports@https://unpkg.com/axios/dist/axios.min.js:8:4321 e.exports/</l[h]@https://unpkg.com/axios/dist/axios.min.js:8:3278` in the browser console. ...

in `layout.pug` `handleFormSubmit` instead of `else` add:
```js
.catch(() => self.createFlash('warning', 'Something went wrong '+msg+'ing your team'))
```

by default `axios` resolves for `status >= 200 && status < 300` and rejects otherwise

**Comment by tbushman** [comment]
@ValeraS I can confirm that was the problem. Thank you! The latest push has this fix.

**Comment by ValeraS** [comment]
@tbushman Could you delete `''` from `list` method in `team.controler.js`. It causes an error when rendering the empty leaderboard..
```js
function list(req, res) {
  return Team.find({})
    .sort()
    .lean()
    .then((teams) => {
      if (teams.length === 0) {
        return res.status(200).json(['']); // <--- here
      }
      return res.status(200).json(teams);
    });
}
```

**Comment by tbushman** [comment]
@raisedadead I am merging this. Thanks @ValeraS for QA.

**Pull request merged**